stdlib/thread_local_storage.mli

```diff
@@ -0,0 +1,74 @@
+(**************************************************************************)
+(*                                                                        *)
+(*                                 OCaml                                  *)
+(*                                                                        *)
+(*      KC Sivaramakrishnan, Indian Institute of Technology, Madras       *)
+(*                 Stephen Dolan, University of Cambridge                 *)
+(*                   Tom Kelly, OCaml Labs Consultancy                    *)
+(*                                                                        *)
+(*   Copyright 2019 Indian Institute of Technology, Madras                *)
+(*   Copyright 2014 University of Cambridge                               *)
+(*   Copyright 2021 OCaml Labs Consultancy Ltd                            *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+(** Thread-local Storage.
+
+    @since 5.2 *)
+
+type 'a t
+(** Type of a thread-local storage key *)
+
+val make: ?split_from_parent:('a -> 'a) -> (unit -> 'a) -> 'a t
```

Comment:
@lpw25 I'm not sure what you mean with the `'a` parameter of `make_with_split`. Maybe your intention is that this is the initial value on the main domain, that new domains will split from. But note that at the point where we call `make_with_split`, there may be several domains running already, what is the behavior you have in mind in this case?

With the current API, any domain that was created before the key will call the `(unit -> 'a)` initializer to get its value, lazily, on the first key access. (Note that this `(unit -> 'a)` thunk could include some subtle logic to derive those values from a common one. The only thing it cannot reconstruct is the parent-child relationship between existing domains.)

Some possible approaches that one may have in mind:
- All existing domains get assigned the same `'a` value. I don't think this works, especially in the case where `'a` has mutable state, as the result could be very racy.
- The initial domain has the '`a` value. Other pre-existing domains initialize their value by splitting the initial-domain value on the first access. This is probably the best we can do with the API you propose, but note that for non-pre-existing domains the split functions is called *on the parent*, not on the child, and here we would call it on the child. I don't think this really works well. (In particular, the split function may mutate the parent's state, this is what the PRNG split does, and calling it from the child with the parent's value is very racy again.)

Reply:
Note: I am reopening the discussion of DLS API in #13356 and would be happy to incorporate @lpw25's suggestion, but I am currently blocked by the interrogations above.

Reply:
I wrote that a while ago, but I think I was suggesting that if you use a `split` that it should run eagerly for existing domains starting from the initial domain based on the tree structure of their creation. This gives you a consistent semantics regardless of when you happen to call `make` relative to the creation of other domains. Perhaps the cost of running the split functions eagerly overrides the benefit, but I would have thought that the cases that wanted `split` were more likely to be run at top-level before the creation of any additional domains.

Separately from that part of the suggestion, I really think you should separate `make_with_split` from `make` without. Providing the split parameter fundamentally changes the behaviour of the function, which is not a good use of optional arguments. Another way to put it is that there is no value for `split_from_parent` that reproduces the behaviour of providing no `split_from_parent` and I think that is an API smell.

Reply:
> t I think I was suggesting that if you use a split that it should run eagerly for existing domains starting from the initial domain based on the tree structure of their creation.

This would be tricky to implement as those existing domains are already running concurrently.

I guess that we could interrupt domains in turn following the spawning tree, so that they call the split function asynchronously at the next poll point. First we interrupt domain 0, it splits a value for each domain that it spawned, and then in turn it interrupts those domains if they themselves have children, etc. We can do this asynchronously, only `DLS.get` operations need to block until the value for the current child domain has been computed.

I'm not sure what to do if a domain in the spawn tree has already terminated. (I don't think we currently impose that child domains terminate before their parents, only that all domains are eventually joined.) We could reasonably decide that its own parent (the grand-parent of the domain we are spawning) is then in charge of computing the split value. (I think the behavior is less well-behaved if the child itself does it.)

Reply:
Sorry, I meant that the call to make would immediately run `split` to get the values for all the existing domains, and then set them. This ensures that the splitting is all going to be thread safe. However, it does mean that `split` will not always be run on the parent thread, so if you use `Thread_local_state.get` from within a `split` function you won't always receive the values for the parent thread. Perhaps that is enough of a downside not to do it this way?

I find it a little difficult to judge the split API without a more precise understanding of the nature of its use cases. Are there use cases for it beyond the splittable PRNG? If not, should we tailor it more specifically to that case?